### PR TITLE
Use more selective imports at module imports

### PR DESCRIPTION
**Pull request by wilzbach**
AFAICT selective imports have been fixed in [2.071.0](http://dlang.org/changelog/2.071.0.html#imports-314) and now enough time has passed down the road.

FWIW I still hope that we get [Dependency-Carrying Declarations (DIP1005)](https://github.com/dlang/DIPs/blob/master/DIPs/DIP1005.md) at some point and can completely get rid of these global imports and make Phobos entirely pay for what you need.

### CODE REVIEW

**Comment by wilzbach** [comment]
:sob: 

```
ggplotd 1.1.3: building configuration "ggplotd-test-default"...
source/ggplotd/colour.d(30,7): Deprecation: catch statement without an exception specification is deprecated; use catch(Throwable) for old behavior
source/ggplotd/stat.d(511,12): Error: module std.typecons import 'Erase' not found
source/ggplotd/geom.d(1044,25): Error: template instance ggplotd.stat.statDensity2D!(MapResult!(__lambda3, Zip!(MapResult!(__lambda1, Result), MapResult!(__lambda2, Result)))) error instantiating
source/ggplotd/example.d(120,9):        instantiated from here: geomDensity2D!(MapResult!(__lambda3, Zip!(MapResult!(__lambda1, Result), MapResult!(__lambda2, Result))))
source/ggplotd/stat.d(511,12): Error: module std.typecons import 'Erase' not found
source/ggplotd/geom.d(1044,25): Error: template instance ggplotd.stat.statDensity2D!(MapResult!(__lambda1, double[][])) error instantiating
source/ggplotd/example.d(234,21):        instantiated from here: geomDensity2D!(MapResult!(__lambda1, double[][]))
source/ggplotd/stat.d(511,12): Error: module std.typecons import 'Erase' not found
source/ggplotd/stat.d(603,28): Error: template instance ggplotd.stat.statDensity2D!(Aes!(double[], "x", double[], "y")) error instantiating
dmd failed with exit code 1.
```

**Comment by MetaLang** [comment]
LGTM. Are these changes related the the ggplotd errors, or did the latest release just break it?

**Comment by wilzbach** [comment]
> Are these changes the cause of the the ggplotd errors, or did the latest release just break it?

I haven't seen this particular ggplotd error on the project tester before, so it's most likely from this PR (I rebased to force a rerun to be sure though).

**Comment by MetaLang** [comment]
Probably because some functions were relying on a combination of nested imports and module-level blanket ones, which changing to selective broke.

**Comment by wilzbach** [comment]
> Probably because some functions were relying on a combination of nested imports and module-level blanket ones, which changing to selective broke.

And I foolishly assumed that this was fixed / deprecated, but it seems like it's not:

```d
module foo;
import c;
//import c : NotErase; // <- breaks build of module `bar`
```

```d
module bar;
unittest
{
    import foo : Erase;
    assert(Erase == 2);
}
```

```d
module c;
int Erase = 2;
int NotErase = 2;
```

**Comment by JackStouffer** [comment]
Wait, what? So the `import c;` is somehow acting as a public import?

Did the fix for 314 make the problem its opposite?

**Comment by wilzbach** [comment]
> Wait, what? So the import c; is somehow acting as a public import?

AFAICT yep :/

> Did the fix for 314 make the problem its opposite?

I think https://github.com/dlang/dmd/pull/5485/files simply didn't touch this. Even 2.062 treats `import c` as a public import.

**Comment by wilzbach** [comment]
>> Wait, what? So the import c; is somehow acting as a public import?
> AFAICT yep :/

-> https://issues.dlang.org/show_bug.cgi?id=17630

**Comment by joakim-noah** [comment]
Can you investigate this a bit more and add that info to the bug report, including that it's been around for a while?  This seems like a fairly important import leak remaining, would be good to explore how large it is.  Ironically, using selective imports at module scope, as you do in this PR, actually plugs the leak, but breaks code that unintentionally relies on leaking symbols.

**Comment by wilzbach** [comment]
> Can you investigate this a bit more and add that info to the bug report, including that it's been around for a while?

Yeah [I tried to](https://issues.dlang.org/show_bug.cgi?id=17630). My DMD foo is pretty low atm unfortunately. So I can only say that the leaked symbols are [found in the local `symtab` table](https://github.com/dlang/dmd/blob/master/src/ddmd/dsymbol.d#L1306).

> This seems like a fairly important import leak remaining, would be good to explore how large it is. 

It's _pretty_ large. AFAICT all module-level imports, non-selective imports leak their symbols.
However, even selective imports leak their selected symbols:

```d
import imports.test17630b; // works __falsely__ as public import
import imports.test17630b : Erase; // works __falsely__ as public import
```

Funnily even `private import X;` doesn't work.

**Comment by joakim-noah** [comment]
It seems like the entire implementation of selective imports never took visibility into account properly.  This is where having so few compiler devs' eyes on the source of such a massive language is hurting D.

**Comment by MartinNowak** [comment]
I plan to remove the current visibility deprecation for 2.077, would be a good timing for this change as well. Still need to check how many projects still haven't fixed their deprecations though.

**Comment by joakim-noah** [comment]
Likely just needs to be rebased and updated, would be good if you added Andrei's commented-out selective imports from #5948 too.

File: std/algorithm/comparison.d [hunk 1]
```diff
@@ -59,12 +59,12 @@ T2=$(TR $(TDNW $(LREF $1)) $(TD $+))
 module std.algorithm.comparison;
 
 // FIXME
-import std.functional; // : unaryFun, binaryFun;
+import std.functional : unaryFun, binaryFun;
 import std.range.primitives;
 import std.traits;
 // FIXME
```
Comment (joakim-noah) [inline std/algorithm/comparison.d]: Can be removed?

File: std/algorithm/iteration.d [hunk 2]
```diff
@@ -67,7 +67,7 @@ T2=$(TR $(TDNW $(LREF $1)) $(TD $+))
 module std.algorithm.iteration;
 
 // FIXME
```
Comment (joakim-noah) [inline std/algorithm/iteration.d]: Same here.

File: std/algorithm/mutation.d [hunk 3]
```diff
@@ -80,7 +80,7 @@ module std.algorithm.mutation;
 import std.range.primitives;
 import std.traits : isArray, isBlitAssignable, isNarrowString, Unqual, isSomeChar;
 // FIXME
```
Comment (joakim-noah) [inline std/algorithm/mutation.d]: ditto

File: std/algorithm/searching.d [hunk 4]
```diff
@@ -106,11 +106,11 @@ T2=$(TR $(TDNW $(LREF $1)) $(TD $+))
 module std.algorithm.searching;
 
 // FIXME
```
Comment (joakim-noah) [inline std/algorithm/searching.d]: I'll stop here, just remove all unneeded comments.

**Comment by wilzbach** [comment]
> Oh, I guess you're waiting on the 1-year deprecation period for the selective import leak to pass before getting this in?

Yes and it hasn't even started yet, see e.g. https://github.com/dlang/dmd/pull/6676

I will close this PR now to avoid other people bumping into it, but once the selective import leak has been deprecated, imports at Phobos can finally made selective.
Also as a note to reviewers: new top-level imports can and should always be selective - there's no need to increase the area of the symbol leakage (and it's also dangerous because we could introduce a new symbol leak during the deprecation period).

**Pull request closed without merging**